### PR TITLE
LL-6664  Button spinner

### PR DESCRIPTION
**Pull request by nparigi-ledger**
Add a spinner to existing Button component that appear when button is pressed and onPress function is a Promise.

https://user-images.githubusercontent.com/89014981/130989447-bb19f7a6-d46c-4bfa-83df-42fa455cb173.mp4

### Type

Rebranding component

### Context

[LL-6664]

### Parts of the app affected / Test plan

Uilib / storybook

[LL-6664]: https://ledgerhq.atlassian.net/browse/LL-6664

### CODE REVIEW

**Comment by nparigi-ledger** [comment]
> could you separate this in a new component like PromiseButton or something i don't want to have the logic of the spinner all over the app on, places where it should not occur

Or maybe an option to disable this behavior and logic, true by default ?
After all, it's still a button and need all the same props and params of Button.

**Comment by LFBarreto** [comment]
> > could you separate this in a new component like PromiseButton or something i don't want to have the logic of the spinner all over the app on, places where it should not occur
> 
> Or maybe an option to disable this behavior and logic, true by default ?
> After all, it's still a button and need all the same props, styles and params of Button.

Just be careful not to have Animated nodes that are passively computing stuff even if unused, its a terrible performance issue to have on a widely used component cause it scales with the usages in the pages.
Either you split it or you do a split in an entry component for the logic to not even be on the base button :shrug:

File: uilib/src/components/Button/index.tsx [hunk 1]
```diff
@@ -52,32 +54,87 @@ export const Base = styled.TouchableOpacity<{
...
   ...props
 }: Props): React.ReactElement => {
+  const [spinnerOn, setSpinnerOn] = useState(false);
   const theme = useTheme();
   const { color } = getButtonColor({ ...props, theme });
+
+  const onPressHandler = useCallback(async () => {
+    if (!onPress) return;
+    let isPromise;
+    try {
```
[the remark this answers is not in the document]
Comment (nparigi-ledger) [reply uilib/src/components/Button/index.tsx:136]: I have now simplified this function a lot, what do you think about it ?

**Comment by nparigi-ledger** [comment]
> > > could you separate this in a new component like PromiseButton or something i don't want to have the logic of the spinner all over the app on, places where it should not occur
> > 
> > 
> > Or maybe an option to disable this behavior and logic, true by default ?
> > After all, it's still a button and need all the same props, styles and params of Button.
> 
> Just be careful not to have Animated nodes that are passively computing stuff even if unused, its a terrible performance issue to have on a widely used component cause it scales with the usages in the pages.
> Either you split it or you do a split in an entry component for the logic to not even be on the base button

I have now splitted the spinner button to a special "PromisableButton"

**Review by LFBarreto — APPROVED** [review]

**Comment by LFBarreto** [comment]
closing this as it has been included in [ledgerHQ/ui](https://github.com/LedgerHQ/ui/tree/main/packages/native)

**Pull request closed without merging**